crates/subspace-networking/src/peer_info.rs

```diff
@@ -0,0 +1,207 @@
+mod handler;
+mod protocol;
+
+use crate::peer_info::handler::HandlerInEvent;
+use handler::Handler;
+pub use handler::{Config, PeerInfoError, PeerInfoSuccess};
+use libp2p::core::{Endpoint, Multiaddr};
+use libp2p::swarm::behaviour::{ConnectionEstablished, FromSwarm};
+use libp2p::swarm::{
+    ConnectionDenied, ConnectionId, NetworkBehaviour, NotifyHandler, PollParameters, THandler,
+    THandlerInEvent, THandlerOutEvent, ToSwarm,
+};
+use libp2p::PeerId;
+use parity_scale_codec::{Decode, Encode};
+use std::collections::VecDeque;
+use std::task::{Context, Poll};
+use tracing::debug;
+
+#[derive(Clone, Debug, Encode, Decode)]
+/// Peer info data
+pub struct PeerInfo {
+    pub role: PeerRole,
+}
+
+impl Default for PeerInfo {
+    fn default() -> Self {
+        PeerInfo {
+            role: PeerRole::Client,
+        }
+    }
+}
+
+#[derive(Clone, Debug, Encode, Decode)]
+/// Defines the role of a peer
+pub enum PeerRole {
+    Farmer,
+    Node,
+    BootstrapNode,
+    Client,
+}
+
+/// The result of an inbound or outbound peer-info requests.
+pub type Result = std::result::Result<PeerInfoSuccess, PeerInfoError>;
+
+/// A [`NetworkBehaviour`] that handles inbound peer info requests and
+/// sends outbound peer info requests on the first established connection.
+pub struct Behaviour<PeerInfoProvider = ConstantPeerInfoProvider> {
+    /// Peer info protocol configuration.
+    config: Config,
+    /// Queue of events to yield to the swarm.
+    events: VecDeque<Event>,
+    /// Outbound peer info pushes.
+    requests: Vec<Request>,
+    /// Provides up-to-date peer info.
+    peer_info_provider: PeerInfoProvider,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+/// Peer info push request.
+struct Request {
+    peer_id: PeerId,
+}
+
+/// Provides the current peer info data.
+pub trait PeerInfoProvider: 'static {
+    /// Returns the current peer info data.
+    fn peer_info(&self) -> PeerInfo;
+}
+
+/// Handles constant peer info data.
+pub struct ConstantPeerInfoProvider {
```

Comment:
I didn't intend to do that. It was meant for "Client" or "Bootstrap node" peers. It could be renamed though if you suggest better name.

Reply:
I'm confused, by the last commits of this PR you have already added `data` field to `PeerInfo` and later updates of cuckoo filter, there doesn't seem to be anything "constant" to it unless I'm missing something.

Reply:
Hm, I see, you have two implementations. But you configure networking with just one of them, it doesn't seem to be possible to use both at the same time. Why having `ConstantPeerInfoProvider` at all in that case?

Reply:
If we encode a message on the wire, decoding side should use exactly the same data structure to be able to decode it. If some nodes on the network use `ConstantPeerInfoProvider` and others use `ArchivalStoragePieces`, how would they even talk to each other? My understanding is that everyone should use the same generics to be compatible.

Reply:
Sorry, I still don't see your point. "Message-on-the-wire" is the same for all peers and it is `PeerInfo`, encoding and decoding are also the same and it is SCALE. Different structures provide `PeerInfo` for different peers - that is correct. What issue do you observe?

Reply:
I think I now understand what you're saying. Now it seems that `ConstantPeerInfoProvider` is almost identical to `ArchivalStoragePieces` except it is missing said pieces and as the result doesn't fire any events.

What if we have just one `PeerInfoProvider` enum, no traits. It will have several constructors, one for each peer info type. Constructor for farmer will take `ArchivalStoragePieces` as an argument.

That way it is all in one place and `ArchivalStoragePieces` is clearly separated (maybe behind a trait, but probably not necessary initially) as an external source of information.